core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -32,104 +58,126 @@
-        this.name = name;
-        this.values = values;
+    public TermsQueryBuilder(String fieldName, float... values) {
+        this(fieldName, Floats.asList(values));
     }
 
     /**
      * A filter for a field based on several terms matching on any of them.
      *
-     * @param name   The field name
+     * @param fieldName The field name
      * @param values The terms
      */
-    public TermsQueryBuilder(String name, double... values) {
-        this.name = name;
-        this.values = values;
+    public TermsQueryBuilder(String fieldName, double... values) {
+        this(fieldName, Doubles.asList(values));
     }
 
     /**
      * A filter for a field based on several terms matching on any of them.
      *
-     * @param name   The field name
+     * @param fieldName The field name
      * @param values The terms
      */
-    public TermsQueryBuilder(String name, Object... values) {
-        this.name = name;
-        this.values = values;
+    public TermsQueryBuilder(String fieldName, Object... values) {
+        this(fieldName, values != null ? Arrays.asList(values) : null);
+    }
+
+    /**
+     * Constructor used for terms query lookup.
+     *
+     * @param fieldName The field name
+     */
+    public TermsQueryBuilder(String fieldName) {
+        this(fieldName, (Object[]) null);
     }
 
     /**
      * A filter for a field based on several terms matching on any of them.
      *
-     * @param name   The field name
+     * @param fieldName The field name
      * @param values The terms
      */
-    public TermsQueryBuilder(String name, Iterable values) {
-        this.name = name;
-        this.values = values;
+    public TermsQueryBuilder(String fieldName, Iterable values) {
+        this.fieldName = fieldName;
+        if (values != null) {
+            this.values = convertToBytesRefListIfStringList(values);
+        }
+        isTermsLookupQuery(values == null);
+    }
+
+    public String fieldName() {
+        return this.fieldName;
+    }
+
+    public List<Object> values() {
+        return convertToStringListIfBytesRefList(this.values);
+    }
+
+    public TermsQueryBuilder isFilter(boolean isFilter) {
+        this.isFilter = isFilter;
+        return this;
+    }
+    
+    public boolean isFilter() {
+        return this.isFilter;
     }
 
     /**
-     * Sets the execution mode for the terms filter. Cane be either "plain", "bool"
-     * "and". Defaults to "plain".
-     * @deprecated elasticsearch now makes better decisions on its own
+     * Sets whether the query is a terms lookup query.
      */
-    @Deprecated
-    public TermsQueryBuilder execution(String execution) {
-        this.execution = execution;
-        return this;
+    private void isTermsLookupQuery(boolean isTermLookupQuery) {
+        this.isTermsLookupQuery = isTermLookupQuery;
+    }
+
+    public boolean isTermsLookupQuery() {
+        return this.isTermsLookupQuery;
     }
```

Comment:
I think that we can do without this flag, it could be a util method that returns a boolean depending on whether values are set or lookup details are.